File: app/src/main/java/com/androidstudy/mpesa/common/DarajaLiveData.java
package com.androidstudy.mpesa.common;

import android.arch.lifecycle.LiveData;
import android.support.annotation.NonNull;

import com.androidstudy.daraja.DarajaListener;


public class DarajaLiveData<T> extends LiveData<Resource<T>> implements DarajaListener<T> {

    public DarajaLiveData() {
        setValue(new Resource<>(Status.LOADING));
    }

    @Override
    public void onResult(@NonNull T data) {
        setValue(new Resource<T>(data));
    }

    @Override
    public void onError(String error) {
        setValue(new Resource<T>(new DarajaException(error)));
    }
}

File: app/src/main/java/com/androidstudy/mpesa/common/DarajaLiveData.kt
/*
 * Copyright 2021 Juma Allan
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.androidstudy.mpesa.common

import androidx.lifecycle.LiveData
import com.androidstudy.daraja.callback.DarajaException
import com.androidstudy.daraja.callback.DarajaListener

class DarajaLiveData<T> : LiveData<Resource<T>>(), DarajaListener<T> {

    init {
        value = Resource(Status.LOADING)
    }

    override fun onResult(result: T) {
        value = Resource(result)
    }

    override fun onError(exception: DarajaException) {
        value = Resource(DarajaException(exception.localizedMessage))
    }
}